indy_node/server/action_log.py

```diff
@@ -1,62 +1,196 @@
+
+
+# Note. datetime class is referred not directly (through datetime module)
+# since it makes possible to mock it in tests using subclasses
+class ActionLogEvent(CsvSerializer):
+    def __init__(
+            self,
+            ts: Union[datetime.datetime, str],
+            ev_type: Enum,
+            data: Type[Union[CsvSerializer, str]],
+            *args: Tuple[str],
+            data_class: Type[Union[CsvSerializer, None]] = None,
+            types: Type[Enum] = ActionLogEvents
+    ):
+        if ts:
+            if isinstance(ts, str):
+                ts = parse_dt(ts)
+            if not isinstance(ts, datetime.datetime):
+                raise TypeError(
+                    "'ts' should be 'datetime' or None, got: {}"
+                    .format(type(ts))
+                )
+
+        if isinstance(ev_type, str):
+            try:
+                ev_type = types[ev_type]
+            except KeyError:
+                raise ValueError("Unknown event {}"
+                                 .format(ev_type))
+
+        if not isinstance(ev_type, Enum):
+            raise TypeError(
+                "'ev_type' should be 'Enum', got: {}"
+                .format(type(ts))
+            )
+
+        if ev_type not in types:
+            raise ValueError("Unknown event {}".format(ev_type))
+
+        data = data_class(data, *args) if data_class else data
+        if not isinstance(data, CsvSerializer):
+            raise TypeError(
+                "'data' should be 'CsvSerializer', got: {}"
+                .format(type(data))
+            )
+
+        self.ts = ts if ts else datetime.datetime.utcnow()
+        self.ev_type = ev_type
+        self.data = data
+
+        self._data_items_prefix = '_data_'
+        self._items = (
+            ['ts', 'ev_type'] +
+            [(self._data_items_prefix + i) for i in self.data._items]
+        )
+        pass
+
+    def __getattr__(self, name):
+        try:
+            _name = name.split(self._data_items_prefix)[1]
+        except IndexError:
+            raise AttributeError
+        else:
+            return getattr(self.data, _name)
+
+
+class ActionLog:
     """
     Append-only event log of action event
     """
 
-    SCHEDULED = "scheduled"
-    STARTED = "started"
-    SUCCEEDED = "succeeded"
-    FAILED = "failed"
-    CANCELLED = "cancelled"
-
-    def __init__(self, filePath, delimiter="\t"):
-        self.__delimiter = delimiter
-        self.__filePath = filePath
-        self.__items = []
+    def __init__(
+            self,
+            file_path: str,
+            data_class: Type[CsvSerializer] = ActionLogData,
+            event_types: Type[Enum] = ActionLogEvents,
+            delimiter: str ='\t'
+    ):
+        self._delimiter = delimiter
+        self._file_path = file_path
+        self._items = []
+        self._data_class = data_class
+        self._event_types = event_types
         self._load()
 
+        for ev_type in self._event_types:
```

Comment:
explicit methods here can't satisfy the constructor API since it accepts any Enum, as a trade-off I can add explicit ones in child classes where enum for types becomes defined

Reply:
done